Partial Data State

kuasha420 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8141

## Relevant technical choices

This is mostly implemented as IB. I added one additional selector `haveAudiences` following the prior art from `haveCustomDimensions` for validation. 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.js:
```diff
@@ -0,0 +1,578 @@
+				propertyID,
+				RESOURCE_TYPE_PROPERTY
+			)
+	),
+	/**
```
```suggestion
	),

	/**
```

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.js:
```diff
@@ -0,0 +1,578 @@
+			switch ( resourceType ) {
+				case RESOURCE_TYPE_AUDIENCE:
+					return {
+						startDate: getDateString(
+							new Date( propertyCreateTime )
+						),
```
As this is duplicated across all types, how about we create a `startDate` variable above and just use that everywhere?

nfmohit commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -130,6 +132,42 @@ const baseResolvers = {
+	 * @param {string|Array<string>} audienceResourceNames Custom dimensions to check.
+	 * @return {boolean} True if all provided audiences are available, otherwise false. Undefined if available audiences are not loaded yet.
+	 */
+	haveAudiences: createRegistrySelector(
```
Does `hasAudiences` sound more grammatically correct, as audiences belong to a singular property entity?

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.js:
```diff
@@ -0,0 +1,578 @@
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+import { createReducer } from '../../../googlesitekit/data/create-reducer';
+import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
+import { getDateString } from '../../../util';
+import { DATE_RANGE_OFFSET, MODULES_ANALYTICS_4 } from './constants';
```
Should we improve the order of these imports?

```suggestion
import API from 'googlesitekit-api';
import Data from 'googlesitekit-data';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { createReducer } from '../../../googlesitekit/data/create-reducer';
import { getDateString } from '../../../util';
import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
import { DATE_RANGE_OFFSET, MODULES_ANALYTICS_4 } from './constants';
```

kuasha420 replied:
That's how VSCode orders the import, I think it takes path into consideration first before the imported symbol.

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.test.js:
```diff
@@ -0,0 +1,730 @@
+/**
+ * Internal dependencies
+ */
+import {
+	createTestRegistry,
+	provideModules,
+	provideUserAuthentication,
+	untilResolved,
+} from '../../../../../tests/js/utils';
+import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
+import { getPreviousDate, stringToDate } from '../../../util';
+import { properties } from './__fixtures__';
+import { DATE_RANGE_OFFSET, MODULES_ANALYTICS_4 } from './constants';
+import { RESOURCE_TYPE_AUDIENCE } from './partial-data';
```
Should we improve the order of these imports?

```suggestion
import {
	createTestRegistry,
	provideModules,
	provideUserAuthentication,
	untilResolved,
} from '../../../../../tests/js/utils';
import { getPreviousDate, stringToDate } from '../../../util';
import { properties } from './__fixtures__';
import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
import { DATE_RANGE_OFFSET, MODULES_ANALYTICS_4 } from './constants';
import { RESOURCE_TYPE_AUDIENCE } from './partial-data';
```

kuasha420 replied:
Same as above. 

> That's how VSCode orders the import, I think it takes path into consideration first before the imported symbol.

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.test.js:
```diff
@@ -0,0 +1,730 @@
+	'^/google-site-kit/v1/modules/analytics-4/data/save-resource-data-availability-date'
+);
+
+// TODO: use `replaceAll` instead of `replace` accross the file when we upgrade our Node version.
```
```suggestion
// TODO: use `replaceAll` instead of `replace` across the file when we upgrade our Node version.
```

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.test.js:
```diff
@@ -0,0 +1,730 @@
+				).toEqual( 20201220 );
+			} );
+
+			it( 'uses a resolver to determine the date and if the date is available. save it to server', async () => {
```
```suggestion
			it( 'uses a resolver to determine the date and if the date is available, saves it to server', async () => {
```

nfmohit reviewed (changes requested):
Excellent work here, thank you @kuasha420. I've left my CR feedback. Please let me know what you think, thank you!

kuasha420 commented on includes/Modules/Analytics_4/Resource_Data_Availability_Date.php:
```diff
@@ -0,0 +1,249 @@
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	const CUSTOM_DIMENSION_SLUGS = array(
+		'googlesitekit_post_type',
+	);
```
This is the only custom dimension relevant for Audience stuff. We can add others if they become relevant.

kuasha420 commented on includes/Modules/Analytics_4/Resource_Data_Availability_Date.php:
```diff
@@ -0,0 +1,249 @@
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param array/null  $available_audience_names Optional. List of available audience resource names. If not provided, it will be fetched from settings.
+	 * @param string/null $property_id Optional. Property ID. If not provided, it will be fetched from settings.
+	 */
+	public function reset_all_resource_dates( $available_audience_names = null, $property_id = null ) {
+
+		foreach ( self::CUSTOM_DIMENSION_SLUGS as $custom_dimension ) {
+			$this->reset_resource_date( $custom_dimension, self::RESOURCE_TYPE_CUSTOM_DIMENSION );
+		}
```
The custom dimensions are statically defined in code, but the others are pulled from settings. The optional parameters are only relevant when we can't guarantee the stability of settings. ie. on `on_change` method.

kuasha420 commented on tests/phpunit/integration/Modules/Analytics_4Test.php:
Great point, added additional test coverage for the items you've mentioned. The code for this is kind of sketchy as the `$resource_data_availability_date` property is not public for `Analytics_4` class. But it should do for test. Thank you!

kuasha420 commented:
@nfmohit Thank you for your thorough review here. I've addressed all your feedback and answered the questions you raised. 

Additionally, we missed a case where the resource available date for audiences that become unavailable weren't being reset. This wasn't explicitly mentioned in AC or IB, but should be done. See the commit [here](https://github.com/google/site-kit-wp/pull/8601/commits/f943ddd85b87244dafa719cd34582fb8c54c620f). 

Back to you for another pass. 

Cheers.

nfmohit commented on assets/js/modules/analytics-4/datastore/partial-data.test.js:
```diff
@@ -237,7 +237,7 @@ describe( 'modules/analytics-4 partial data', () => {
 			} );
 
-			it( 'uses a resolver to determine the date and if the date is available. save it to server', async () => {
+			it( 'uses a resolver to determine the date and if the date is available. saves it to server', async () => {
```
I think we'd want to use a comma here instead of a period.

```suggestion
			it( 'uses a resolver to determine the date and if the date is available, saves it to server', async () => {
```

nfmohit reviewed (changes requested):
Thanks so much for addressing all my feedback, @kuasha420. I have left a couple additional comments which are very minor, so minor that I'm going to address them myself. I hope that is okay with you, thanks!

kuasha420 commented on tests/phpunit/integration/Modules/Analytics_4Test.php:
```diff
@@ -65,6 +66,8 @@
 use WP_User;
 use ReflectionMethod;
 
+use function PHPUnit\Framework\assertNotFalse;
```
Ahhh, stupid VSCode extension auto imports all these stuff :(
